docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+- **Allocation of non-compressible resources (e.g. memory) to top-tier pods will be the same as if best-effort pods did not exist**. For example consider a system with 4GB of memory, where top-tier pod P1 is using 500MB of memory (with a 1GB limit) and best-effort pods are using the other 3.5GB of memory. The best-effort pods will not prevent P1 from accessing more memory (assuming that P1 remains within its memory limits). So if P1 asks for 100MB more memory, P1 will get the memory it needs, and one of the other pods will be OOM killed. 
+
+Resource guarantees for best-effort pods:
+- Compressible resources
+  - For now, we are only implementing CPU limits.
+  - CPU shares not used by top-tier jobs will be distributed proportionally amongst best-effort pods (based on the amount of CPU the pods requested).
+  - Best effort pods will not be evicted because of CPU constraints, they will simply be starved if CPU resources are not available.
+- Non-compressible resources
+  - For now, we are only implementing memory limits.
+  - We will always launch a best-effort pod, regardless of its memory requirements.
+  - Best effort pods will be evicted if a top tier pod needing the memory requirements comes in, or if some other top-tier pod asks for memory.
+  - Best-effort pods will not be allowed to exceed their memory limit.
+  - Extension: Consider distinguishing between best-effort pods that specify memory limits and give them even lower priority. Otherwise a best-effort pod without a memory limit could use up system resources and kill off other best-effort pods.
+- Extension: support other resources like network, disk.
+
+Resource guarantees for top tier pods
+- Will not change.
+- Extension: Consider giving top-tier pods that do not specify memory limits a lower memory priority than top tier pods that specify memory limits (but a higher priority than best-effort pods)
+
+## Policy
+
+The QoS design does not make assumptions about the “policy” used to label pods as best-effort and top-tier.
+
+In the first iteration, we will stick to a simple policy. Pods that specify CPU and memory limits for every container will be classified as top-tier, all other pods will be classified as best-effort. This policy offers protection to pods that specify resource requirements (from pods that have no limits and can go overboard in resource consumption), which gives them better resource guarantees. 
+
+Extension: In the future, we might add a field to PodSpec in the API, to allow users to specify whether a pod is best-effort or top-tier.

```

Comment:
I'll update the proposal with requests today!

> On 24 Jul, 2015, at 2:31 am, Marek Grabowski <notifications@github.com> wrote:
> 
> In docs/proposals/kubelet-qos.md:
> 
> > +  - Best effort pods will be evicted if a top tier pod needing the memory requirements comes in, or if some other top-tier pod asks for memory.
> > +  - Best-effort pods will not be allowed to exceed their memory limit.
> > +  - Extension: Consider distinguishing between best-effort pods that specify memory limits and give them even lower priority. Otherwise a best-effort pod without a memory limit could use up system resources and kill off other best-effort pods.
> > +- Extension: support other resources like network, disk.
> > +
> > +Resource guarantees for top tier pods
> > +- Will not change.
> > +- Extension: Consider giving top-tier pods that do not specify memory limits a lower memory priority than top tier pods that specify memory limits (but a higher priority than best-effort pods)
> > +
> > +## Policy
> > +
> > +The QoS design does not make assumptions about the “policy” used to label pods as best-effort and top-tier.
> > +
> > +In the first iteration, we will stick to a simple policy. Pods that specify CPU and memory limits for every container will be classified as top-tier, all other pods will be classified as best-effort. This policy offers protection to pods that specify resource requirements (from pods that have no limits and can go overboard in resource consumption), which gives them better resource guarantees. 
> > +
> > +Extension: In the future, we might add a field to PodSpec in the API, to allow users to specify whether a pod is best-effort or top-tier.
> You're right of course: "request" not "requirement" - too many similar words around:/ The thing is I don't see the proposal on how to incorporate requests here, and I think this should be a prerequisite to any QoS work.
> 
> —
> Reply to this email directly or view it on GitHub.
>